telemetry/convert.py

```diff
@@ -146,10 +149,127 @@ def convert_json(self, jsonstr, date, ip=None):
+                   val = val[field]
+                else:
+                    # this unified ping doesn't report this particular field
+                    val = None
+                    break
+            # Copy the value into ping.info[field name]
+            if val != None:
+                info[dstField] = val
+
+    def convert_saved_session(self, json_dict):
+        # Step 1 of conversion:
+        #   ping.payload.* => ping.* (histograms, info, etc)
+        payload = json_dict["payload"]
+        for field in payload:
+            if field == "info" or field == "ver":
+                # info field needs to be merged, ping.ver has different meaning
+                continue
+            elif field in json_dict:
+                # TODO: Make this throw?
+                json_dict["payload." + str(field)] = payload[field]
+                continue
+            else:
+                json_dict[field] = payload[field]
+
+        # Merge ping.payload.info.* fields to ping.info.*
+        payload_info = payload["info"]
+        for field in payload_info:
+            if field not in json_dict["info"]:
+                json_dict["info"][field] = payload_info[field]
+        # Back up payload.ver
+        json_dict["payload.ver"] = payload["ver"]
+
+        # Get rid of duplicated data in the payload field
+        del json_dict["payload"]
+
+        # Step 2 of conversion:
+        #   Recreate the old-style ping.info section from fields that are
+        #   now in ping.environment
+        envFields = json_dict["environment"]
+        info = json_dict["info"]
+        self.add_info_fields(info, envFields, envFieldMap)
+
+        adapters = None
+        try:
+            adapters = envFields["system"]["gfx"]["adapters"]
+        except KeyError, TypeError:
+            pass
+
+        if type(adapters) == list and len(adapters):
+            self.add_info_fields(info, adapters[0], adapterFieldMap)
+            if len(adapters) > 1:
+                # Copy details of the second GPU
+                self.add_info_fields(info, adapters[1], adapterFieldMap, "2")
+                info["isGPU2Active"] = bool(adapters[1].get("GPUActive"))
+
+        # Fix up clientID
+        json_dict["clientID"] = json_dict.get("clientId")
+
+        # Step 3:
+        #   Add a "ver" field to make it look like a classic ping, and
+        #   delete the "version" field to make the conversion idempotent.
+        #   This ping will then be passed through convert_obj again to
+        #   fix up the histograms.
+        json_dict["original.version"] = json_dict["version"]
+        del json_dict["version"]
+        json_dict["ver"] = Converter.VERSION_UNCONVERTED
+
     def convert_obj(self, json_dict, date, ip=None):
-        info = json_dict.get("info", None)
-        # Check if the payload is already converted:
-        if "ver" in json_dict:
+        if "version" in json_dict:
+            # This is a new unified (FHR+Telemetry) ping
```

Comment:
Add a link to the spec https://ci.mozilla.org/job/mozilla-central-docs/Tree_Documentation/toolkit/components/telemetry/telemetry/index.html